### ISSUE TITLE
Memory leak `java.lang.OutOfMemoryError`: Direct buffer memory (due to timeouts?)

### ISSUE THREAD

**Issue by jensbaitingerbosch** [promregator/promregator#115]
## Environment
cf api: 2.120.0
promregator: 0.5.7 (also tried 0.6.0-SNAPSHOT)

## Summary / Problem Statement
Currently our CF api is very slow (due to updates and other problems). During this time, the discovery fails from time to time with timeouts:

(stack traces in the wrong order, I dont know why that happens)
```
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_212]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_212]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_212]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[na:1.8.0_212]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_212]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_212]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:370) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:253) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:268) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 2500ms (and no fallback has been configured)
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] 2019-07-31 06:55:12.612 WARN 14 --- [ parallel-12] o.c.p.cfaccessor.ReactiveCFAccessorImpl : Woof woof! It appears that the connection to the Cloud Controller is gone. Trying to restart Cloud Foundry Client
```

```
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_212]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_212]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_212]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[na:1.8.0_212]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_212]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_212]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:370) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:253) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:268) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 10000ms (and no fallback has been configured)
```

During that time the following error occured after some time: 
```
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:27) [reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:50) [reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:370) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:253) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:268) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.SerializedSubscriber.onError(SerializedSubscriber.java:114) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxPeek$PeekSubscriber.onError(FluxPeek.java:198) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at org.cloudfoundry.promregator.cfaccessor.ReactiveCFAccessorImpl.lambda$connectionWatchdog$0(ReactiveCFAccessorImpl.java:175) ~[classes/:na]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at org.cloudfoundry.promregator.cfaccessor.ReactiveCFAccessorImpl.constructCloudFoundryClient(ReactiveCFAccessorImpl.java:159) ~[classes/:na]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Mono.block(Mono.java:1182) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoOnAssembly.subscribe(MonoOnAssembly.java:76) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoPeek.subscribe(MonoPeek.java:71) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoNext.subscribe(MonoNext.java:40) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoFlatMapMany.subscribe(MonoFlatMapMany.java:49) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoFlatMap.subscribe(MonoFlatMap.java:60) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoMap.subscribe(MonoMap.java:55) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoProcessor.subscribe(MonoProcessor.java:444) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoProcessor.add(MonoProcessor.java:531) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Mono.subscribe(Mono.java:3088) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelayUntil.subscribe(MonoDelayUntil.java:83) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoOnErrorResume.subscribe(MonoOnErrorResume.java:44) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoMapFuseable.subscribe(MonoMapFuseable.java:59) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoSwitchIfEmpty.subscribe(MonoSwitchIfEmpty.java:44) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoJust.subscribe(MonoJust.java:54) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.onSubscribe(MonoDelayUntil.java:122) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:1640) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.onNext(MonoDelayUntil.java:132) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.subscribeNextTrigger(MonoDelayUntil.java:173) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:45) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Operators.complete(Operators.java:128) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelayUntil$DelayUntilTrigger.onComplete(MonoDelayUntil.java:290) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.signal(MonoDelayUntil.java:211) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1083) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:108) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoProcessor.onNext(MonoProcessor.java:389) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.ipc.netty.http.client.MonoHttpClientResponse.subscribe(MonoHttpClientResponse.java:83) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoRetryPredicate.subscribe(MonoRetryPredicate.java:51) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxRetryPredicate$RetryPredicateSubscriber.resubscribe(FluxRetryPredicate.java:123) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoCreate.subscribe(MonoCreate.java:54) ~[reactor-core-3.1.12.RELEASE.jar!/:3.1.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.ipc.netty.tcp.TcpClient.lambda$newHandler$2(TcpClient.java:188) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.ipc.netty.resources.DefaultLoopResources.onClient(DefaultLoopResources.java:183) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.ipc.netty.resources.DefaultLoopResources.cacheNativeClientLoops(DefaultLoopResources.java:247) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.ipc.netty.resources.DefaultLoopResources.cacheNativeServerLoops(DefaultLoopResources.java:233) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at reactor.ipc.netty.resources.DefaultLoopEpoll.newEventLoopGroup(DefaultLoopEpoll.java:64) ~[reactor-netty-0.7.12.RELEASE.jar!/:0.7.12.RELEASE]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:68) ~[netty-transport-native-epoll-4.1.31.Final-linux-x86_64.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:91) ~[netty-transport-native-epoll-4.1.31.Final-linux-x86_64.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:104) ~[netty-transport-native-epoll-4.1.31.Final-linux-x86_64.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:59) ~[netty-transport-4.1.31.Final.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:47) ~[netty-common-4.1.31.Final.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:58) ~[netty-common-4.1.31.Final.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:84) ~[netty-common-4.1.31.Final.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.EpollEventLoopGroup.newChild(EpollEventLoopGroup.java:35) ~[netty-transport-native-epoll-4.1.31.Final-linux-x86_64.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.EpollEventLoopGroup.newChild(EpollEventLoopGroup.java:135) ~[netty-transport-native-epoll-4.1.31.Final-linux-x86_64.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.EpollEventLoop.<init>(EpollEventLoop.java:90) ~[netty-transport-native-epoll-4.1.31.Final-linux-x86_64.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.EpollEventArray.<init>(EpollEventArray.java:56) ~[netty-transport-native-epoll-4.1.31.Final-linux-x86_64.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.unix.Buffer.allocateDirectWithNativeOrder(Buffer.java:40) ~[netty-transport-native-unix-common-4.1.31.Final.jar!/:4.1.31.Final]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:311) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:123) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] at java.nio.Bits.reserveMemory(Bits.java:694) ~[na:1.8.0_212]
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] java.lang.OutOfMemoryError: Direct buffer memory
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] 2019-07-31 06:55:15.523 ERROR 14 --- [ parallel-12] reactor.core.scheduler.Schedulers : Scheduler worker in group main failed with an uncaught exception
```
After this, the discovery did not find any apps any longer and returns only the promregator as target.

**Comment by eaglerainbow** [comment]
Hi

Thanks for reporting your issue!

Well, it seems that it's even worse than that:

``` 
2019-07-31T08:55:12.612+02:00 [APP/PROC/WEB/0] [OUT] 2019-07-31 06:55:12.612 WARN 14 --- [ parallel-12] o.c.p.cfaccessor.ReactiveCFAccessorImpl : Woof woof! It appears that the connection to the Cloud Controller is gone. Trying to restart Cloud Foundry Client
``` 

indicates that the connection to your CFCC is even broken - it does not respond to the "info" query within a reasonable timeout at all. So Promregator thinks that the platform is even gone - and tries to reset the java client. 

``` 
2019-07-31T08:55:15.524+02:00 [APP/PROC/WEB/0] [OUT] java.lang.OutOfMemoryError: Direct buffer memory
```
is a typical consequence that may happen after a while, if the timeouts used per scraping requests is higher than the scraping interval that you chose (e.g. if it takes more than 30 seconds for the timeouts to detect that the thing won't answer, but you query for data every 15 seconds, then you will start piling up connections). The end of the story then is that you have still-open inbound connections, triggering further outbound connections. However, you can't get them closed fast enough. Each inbound and each outbound connection, however, takes a certain amount of memory - both on the Java Heap and on the operating system's internal (I/O) memory pool. The exact error message only depends then on the which of the two memory areas is exhausted earlier. 

> After this, the discovery did not find any apps any longer and returns only the promregator as target.

Well, not surprising: We don't get any fresh metadata anymore from the CFCC. If the cache times out, then the data is displaced. The `/discovery` request then can only tell you, that there is no knowledge available anymore for the targets...

Though I fear that I can't really help here, we could at least try to check the obvious:
* When you start promregator, do you get a warning about timeout thresholds?
* What are the exact timeout values you have configured?
* What is your scraping interval you are enforcing?
* Are you able to still get promregator's internal metrics via its endpoint? If yes, what is the usual latency you observe for the different caches? What is the average latency for metrics fetching at your application endpoints? What is the sum of all these averages? If this sum is higher than your scraping interval, there is no other chance for you to increase the scraping interval...

Hope this helps a little!

**Comment by eaglerainbow** [comment]
BTW: 
``` 
2019-07-31T08:54:31.200+02:00 [APP/PROC/WEB/0] [OUT] java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 10000ms (and no fallback has been configured)
```
uff, that is a very long time for a platform like CF... If you scrape with 15s now, you certainly will have a big problem in any case...

**Comment by jensbaitingerbosch** [comment]
Thanks for the explanation :) 
* I did not notice any warnings (unfortunately the logs are not stored, so I cannot search them)
* After the CF api was slow I increased the timeouts to 10 seconds or even 100 seconds. (for some calls, even the 10 seconds where too less...)
* Scraping is done in 15 seconds interval
* I was able to get the the promregator metrics all the time. They peak up to 7-10 seconds  and then go down to 100-300ms for 5 minutes until the next peak

**Comment by eaglerainbow** [comment]
Then I would keep the timeouts at around 3 seconds or so: During the peak it does not really make sense to me asking for the metrics - they will be not in sync, out of order.
The only thing to minimize the impact would be to increase the thread pool size `promregator.scraping.threads` drastically: ideally, it would have the same (or larger) value as the total number of *instances* you want to scrape. By this, you could get lucky enough to squeeze in those 10 seconds in maximal parallelization to still comply with the 15s metrics interval driven by Prometheus. However, you must keep in mind that the higher you drive this value, the more memory (on Java Heap and Java Thread Stack) you need and the more open connections in parallel you will require (some environments, such as CF may limit this resource and then you may end up in an even worse situation). 

All in all, why do you have such peaks with such a  long runtime to scrape the metrics? To be honest, I would prefer to invest time into handling that question instead of trying to tune Promregator, which just can try to deal with the bad situation. Is it the network? Is it the platform? Is it the application? If it's the latter, then this would worry me most, as you then don't know whether the metrics values are consistent even within one scraping interval... This can lead to very strange data anomalies...

**Comment by jim-mclean** [comment]
FYI, I'm seeing the same issue in a deployment that I have.  We're running with Single Target Scraping.  I am going to up all the cf.request.time to 60s.

Any chance the new cache support in 0.7.1-rc1 would help with this problem?

**Comment by eaglerainbow** [comment]
> Any chance the new cache support in 0.7.1-rc1 would help with this problem?

There is only one possible answer: *maybe* 😀 
*If* you are getting this issue, because the latency of the CFCC responses is high and therefore the number of asynchronous requests start to accumulate, then Caffeine Cache of 0.7.1-rc1 may help to reduce the problematic pooling. 
However, *if* the root cause for your out-of-memory is related to another entity (e.g. data volume of the metric responses you get back from the containers of your apps), then 0.7.1-rc1 will be of no help at all. 

So, in short, it is worth a try. But don't be too surprised, if it does not help.

**Comment by eaglerainbow** [comment]
@checketts found a [major bug in cf-java-client](https://github.com/cloudfoundry/cf-java-client/pull/1029) which caused the direct buffer memory to overflow. A Promregator version using an updated version of cf-java-client is now available with [0.7.3](https://github.com/promregator/promregator/releases/tag/v0.7.3).

@jim-mclean @jensbaitingerbosch Is it possible to test your use cases again with this newer version and report back if that solved it?

**Comment by jim-mclean** [comment]
I will throw this into our development environment and report back in a couple days.  FYI, The way that I mitigated this issue thus far (beyond the tuning recommendations here) was to have CF hit the /discovery endpoint as the healthcheck.  When that locked up it would mark the container unhealthy and spin a new one up.

**Comment by eaglerainbow** [comment]
> was to have CF hit the /discovery endpoint as the healthcheck. 

In a discussion with @warlockza and @checketts we came to a similar conclusion. The approach, however, only works if your endpoint isn't auth-protected (lucky you, if that is the case). 

At any rate, I'd be interested in hearing whether this fix helps to solve your issue as well.
